[MM] Add sm carevout to lowerings

drisspg opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #148849
* __->__ #148793

# Summary

See https://github.com/pytorch/pytorch/issues/145115 for more details. I have been using 
the following to verify, need to figure out how to do proper guarding


This does  do the correct thing if we compile w/ sm carvout already set but since we dont guard on it just yet we dont recompile

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

davidberard98 commented on torch/_inductor/utils.py:
```diff
@@ -1253,10 +1253,16 @@ def is_big_gpu(index_or_device: Union[int, torch.device] = 0) -> bool:
 
+def get_num_sms() -> int:
+    """Handle experimental carveout if set otherwise return hardware SM count"""
+    carveout = torch._C._get_sm_carveout_experimental()
+    return carveout if carveout is not None else get_max_num_sms()
```
could be wrong, but I thought these were the semantics:

```suggestion
    carveout = torch._C._get_sm_carveout_experimental()
    return get_max_num_sms() - (carveout if carveout is not None else 0)
```

drisspg replied:
Ahh you are very right the formatting on this line was very weird for me and got confused: https://github.com/pytorch/pytorch/pull/144974/files#diff-74fcb26047c1df4024105d36ce22a36b77cf8cc93c28631d743e639b3d6066aeR413-R414

drisspg committed:
Update

[ghstack-poisoned]

lw commented on test/inductor/test_max_autotune.py:
```diff
@@ -463,6 +467,78 @@ def addmm(x, a, b):
+                    self.assertEqual(torch._C._get_sm_carveout_experimental(), 0)
+                    compiled_mm(a, b)
+
+                    # Test with carveout 66
```
In my previous tests for fp8 I chose 66 because the tests only ran on H100 which has 132, so I thought that picking half of that would have the highest chances of giving a visible difference.

lw commented on test/inductor/test_max_autotune.py:
```diff
@@ -463,6 +467,78 @@ def addmm(x, a, b):
+                ]
+
+                # We should have at least 4 kernel events (one for each configuration)
+                self.assertGreaterEqual(len(kernel_events), 4)
```
Why not be stricter here and require exactly 4? Do we ever legitimately expect to have more? What would that imply?

lw reviewed (approved):
Thanks!

eellison reviewed (approved):
Is it worth adding a dynamo guard on this in a follow up ?

drisspg committed:
Update on "[MM] Add sm carevout to lowerings"


# Summary

See https://github.com/pytorch/pytorch/issues/145115 for more details. I have been using 
the following to verify, need to figure out how to do proper guarding


This does  do the correct thing if we compile w/ sm carvout already set but since we dont guard on it just yet we dont recompile

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg committed:
Update

[ghstack-poisoned]

drisspg commented:
Alright updated w/ better test. And added todo about guarding. TLDR is that if you compile w/ a given carevout out this will get backed in and will be cached across calls.
